alg/los.cpp

```diff
@@ -0,0 +1,237 @@
+                        0) == CE_None;
+}
+
+// Check a single location is above terrain.
+static bool IsAboveTerrain(const GDALRasterBandH hBand, const int x,
+                           const int y, const double z)
+{
+    double terrainHeight;
+    if (GetElevation(hBand, x, y, terrainHeight))
+    {
+        return z > terrainHeight;
+    }
+    else
+    {
+        return false;
+    }
+}
+
+/************************************************************************/
+/*                        GDALIsLineOfSightVisible()                    */
+/************************************************************************/
+
+/**
+ * Check Line of Sight between two points.
+ * Both input coordinates must be within the raster coordinate bounds.
+ *
+ * This algorithm will check line of sight using a Bresenham algorithm.
+ * https://www.researchgate.net/publication/2411280_Efficient_Line-of-Sight_Algorithms_for_Real_Terrain_Data
+ * Line of sight is computed in raster coordinate space, and thus may not be appropriate.
+ * For example, datasets referenced against geographic coordinate at high latitudes may have issues.
+ *
+ * @param hBand The band to read the DEM data from. This must NOT be null.
+ * 
+ * @param xA The X location (raster column) of the first point to check on the raster.
+ *
+ * @param yA The Y location (raster row) of the first point to check on the raster.
+ * 
+ * @param zA The Z location (height) of the first point to check.
+ * 
+ * @param xB The X location (raster column) of the second point to check on the raster.
+ *
+ * @param yB The Y location (raster row) of the second point to check on the raster.
+ * 
+ * @param zB The Z location (height) of the second point to check.
+ * 
+ * @param papszOptions Options for the line of sight algorithm (currently ignored).
+ * 
+ * @return True if the two points are within Line of Sight.
+ *
+ * @since GDAL 3.9
+ */
+
+bool GDALIsLineOfSightVisible(const GDALRasterBandH hBand, const int xA,
+                              const int yA, const double zA, const int xB,
+                              const int yB, const double zB,
+                              CPL_UNUSED CSLConstList papszOptions)
+{
+    VALIDATE_POINTER1(hBand, "GDALIsLineOfSightVisible", false);
+
+    // Perform a preliminary check of the start and end points.
+    if (!IsAboveTerrain(hBand, xA, yA, zA))
+    {
+        return false;
+    }
+    if (!IsAboveTerrain(hBand, xB, yB, zB))
+    {
+        return false;
+    }
+
+    // If both X and Y are the same, no further checks are needed.
+    if (xA == xB && yA == yB)
+    {
+        return true;
+    }
+
+    // TODO if both X's or Y's are the same, it could be optimized for vertical/horizontal lines.
+
+    // Use an interpolated Z height with 2D bresenham for the remaining cases.
+
+    // Lambda for computing the square of a number
+    auto SQUARE = [](const int n) -> int { return std::pow(n, 2); };
```

Comment:
I'm wondering how smart optimizers are regarding that.... Below is more obvious (and shorter :-)):
```suggestion
    auto SQUARE = [](const int n) -> int { return n * n ; };
```

Reply:
hum, actually we should do all those distance computations on doubles. Because if someone tries that on a raster with 25,000 pixels wide, 25000**2 = 2,500,000,000 which exceeds INT_MAX

Reply:
It might depend what you enable and your compiler. https://stackoverflow.com/questions/2940367/what-is-more-efficient-using-pow-to-square-or-just-multiply-it-with-itself#:~:text=So%20it%20looks%20like%20the,case%20(60%20times%20quicker).

https://baptiste-wicht.com/posts/2017/09/cpp11-performance-tip-update-when-to-use-std-pow.html

I'll take the suggestion for now, but can re-evaluate when benchmarking is set up.

Reply:
Are you saying convert the indices to double before calling SQUARE, and make SQUARE operate on doubles?

Reply:
> Are you saying convert the indices to double before calling SQUARE, and make SQUARE operate on doubles?

yes

Reply:
0a447cb319e6a9f103a0f5a170d6d430119f57da now uses double math for all of GetZValue()

Reply:
I should add a test for large rasters to check nothing overflows...

Reply:
I tried doing a test with a dataset close to overflow size, and even if its a 1xint_max, it can't allocate the memory. The test just hangs forever. If you have a way to create a dataset that large in a test to check overflow, let me know.

Reply:
Adding tests for large datasets is difficult, and we want the test suite to be as fast as possible. So I would put that aside.

Reply:
Alright. Ready for another pass then.